hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/kafka/SqlAvroSchemaEvolutionTest.java

```diff
@@ -0,0 +1,263 @@
+        name = createRandomTopic(1);
+        kafkaTestSupport.setSubjectNameStrategy(name, false, subjectNameStrategy);
+        topicNameStrategy = subjectNameStrategy.equals("TopicNameStrategy");
+        switch (subjectNameStrategy) {
+            case "TopicNameStrategy":       valueSubjectName = name + "-value"; break;
+            case "TopicRecordNameStrategy": valueSubjectName = name + "-jet.sql"; break;
+            case "RecordNameStrategy":      valueSubjectName = "jet.sql";
+        }
+    }
+
+    private KafkaMapping kafkaMapping() {
+        return new KafkaMapping(name).options(
+                OPTION_KEY_FORMAT, AVRO_FORMAT,
+                OPTION_VALUE_FORMAT, AVRO_FORMAT,
+                "bootstrap.servers", kafkaTestSupport.getBrokerConnectionString(),
+                "schema.registry.url", kafkaTestSupport.getSchemaRegistryURI(),
+                "auto.offset.reset", "earliest",
+                "value.subject.name.strategy", "io.confluent.kafka.serializers.subject." + subjectNameStrategy
+        );
+    }
+
+    @Test
+    public void test_autoRegisterSchema() throws SchemaRegistryException {
+        kafkaMapping()
+            .fields("id INT EXTERNAL NAME \"__key.id\"",
+                    "name VARCHAR")
+            .create();
+
+        insertInitialRecordAndAlterSchema();
+
+        if (updateMapping) {
+            kafkaMapping()
+                .fields("id INT EXTERNAL NAME \"__key.id\"",
+                        "name VARCHAR",
+                        "ssn BIGINT")
+                .optionsIf(!topicNameStrategy,
+                           "value.record.name", "jet.sql2")
+                .createOrReplace();
+        }
+
+        insertAndAssertRecords();
+    }
+
+    @Test
+    public void test_useLatestSchema() throws SchemaRegistryException {
+        // create initial schema
+        kafkaTestSupport.registerSchema(name + "-key", ID_SCHEMA);
+        kafkaTestSupport.registerSchema(valueSubjectName, NAME_SCHEMA);
+
+        kafkaMapping()
+            .fields("id INT EXTERNAL NAME \"__key.id\"",
+                    "name VARCHAR")
+            .options("auto.register.schemas", false,
+                     "use.latest.version", true)
+            .create();
+
+        insertInitialRecordAndAlterSchema();
+
+        if (updateMapping) {
+            kafkaMapping()
+                .fields("id INT EXTERNAL NAME \"__key.id\"",
+                        "name VARCHAR",
+                        "ssn BIGINT")
+                .options("auto.register.schemas", false,
+                         "use.latest.version", true)
+                .optionsIf(!topicNameStrategy,
+                           "value.record.name", "jet.sql2")
+                .createOrReplace();
+        }
+
+        if (topicNameStrategy && !updateMapping) {
+            // insert record against mapping's schema
+            assertThatThrownBy(() -> sqlService.execute("INSERT INTO " + name + " VALUES (29, 'Bob')"))
```

Comment:
can you comment more about reason of this error? Inner cause is `Index 1 out of bounds for length 1` very deep. Is that really expected? What produces mismatched indexes?
If that is expected, should `use.latest.version=true` be strongly discouraged: any update in schema may render MAPPING unusable unless it is updated?
BTW, mapping should work when it does not contain all fields from schema (provided that they are optional).

Reply:
`org.apache.avro.generic.GenericDatumWriter#writeRecord` iterates schema fields, not datum fields. That's why when the writer picks up the latest schema (with `ssn`) and SQL provides a datum without `ssn`, the lookup fails and an `ArrayIndexOutOfBoundsException` is thrown. Based on this, we understand that _schema evolution_ is not an _adaption feature_, but rather a smart schema _selection feature_. Other scenarios work _as expected_ because we don't force the writer to use an incompatible schema: `auto.register.schemas` is already designed _not to fail_, `(key|value).schema.id` fixes the initial schema for mapping and finally `[Topic]RecordNameStrategy` creates another subject, which doesn't introduce a _later_ schema that can be used for the mapping.

Nothing is unexpected from the perspective of Avro writers and schema registry, so nothing should be discouraged. However, the explanation above should be included in the documentation with possibly some examples and/or comparison charts (for different subject name strategies and schema selection modes).

Reply:
Confluent Avro uses `ResolvingDecoder` when `AbstractKafkaAvroDeserializer.specificAvroReaderSchema` is set. But this only happens if we decode to `SpecificRecord`, SQL uses `GenericRecord` so it will not use that - that is a pity, because we know the target schema.
`KafkaAvroDeserializer` is just a thin subclass of AbstractKafkaAvroDeserializer, maybe we could configure it with our schema?

Reply:
anyway, that should be a separate task

Reply:
https://hazelcast.atlassian.net/browse/HZ-2750